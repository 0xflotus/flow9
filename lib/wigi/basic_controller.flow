import wigi/editor_types;
import wigi/recursive_tools;


export {
	// Returns the resulting element, and a bool defining whether it was handled by side effect or not.
	// If false, then the change is done by side effects only
	replaceElement(root : WigiElement, path : [int], element : WigiElement) -> Pair<WigiElement, bool>;
	replaceElementInDocument(state : WigiEditorState, path : [int], e : WigiElement) -> void;

	// Some service functions for controller
	hasFixedChildCount(element: WigiElement) -> bool;
	isWigiComap(e: WigiElement) -> bool;

	// Finds element that satisfies given predicate and contains given position `pos`
	findElementContainingPosition(state: WigiEditorState, predicate: (WigiEditorState, [int]) -> bool, pos: [int]) -> Maybe<[int]>;

	isEmptyDeletableRecursive(state: WigiEditorState, path: [int]) -> bool;

	// Scrolls so that the point is visible
	ensurePointVisible(state : WigiEditorState, point : Point) -> void;

}

replaceElementInDocument(state : WigiEditorState, path : [int], e : WigiElement) -> void {
	root = getValue(state.document);
	pair = replaceElement(root, path, e);
	if (pair.second) {
		nextDistinct(state.document, pair.first);
	}
}

replaceElement(root : WigiElement, path : [int], element : WigiElement) -> Pair<WigiElement, bool> {
	if (path == []) Pair(element, true)
	else {
		first = path[0];
		rest = tail(path);
		switch (root : WigiElement) {
			WigiDocument(r, s): {
				ce : Pair<WigiElement, bool> = replaceElement(r, rest, element);
				if (ce.second)
					Pair(cast(WigiDocument(ce.first, s) : WigiDocument -> WigiElement), true)
				else Pair(root, false);
			}
			WigiStory(p, v, s): {
				if (first >= length(^p)) {
					println("Incorrect path in replace");
				}

				step = indexWigiChildren(root, first, root);
				ce = replaceElement(step, rest, element);
				isFormArrayEmpty = ^v == [];
				if (ce.second) {
					p := replace(^p, first, ce.first);

					if (!isFormArrayEmpty) {
						fa0 = ^(v)[0];
						replaceTropicArray(fa0, first, ce.first);
					}
				}
				// There is never any change in the element itself - it is only side effects
				Pair(root, isFormArrayEmpty);
			}
			WigiParagraph(el, s): {
				if (first >= length(el)) {
					println("Incorrect path in replace");
				}

				step = indexWigiChildren(root, first, root);
				ce = replaceElement(step, rest, element);
				if (ce.second) {
					ne = replace(el, first, ce.first);
					Pair(cast(WigiParagraph(ne, s) : WigiParagraph -> WigiElement), true);
				} else {
					Pair(root, false);
				}
			}
			WigiText(txt, style): Pair(element, true);
			WigiEmpty(): Pair(element, true);
			WigiRecursive(el, type, style): {
				step = indexWigiChildren(root, first, root);
				ce = replaceElement(step, rest, element);
				if (ce.second) {
					ne = replace(el, first, ce.first);
					Pair(cast(WigiRecursive(ne, type, style) : WigiRecursive -> WigiElement), true);
				} else {
					Pair(root, false);
				}
			}
			WigiBlock(type, style): Pair(element, true);
		}
	}
}

ensurePointVisible(state : WigiEditorState, point : Point) -> void {
	wh = getValue(state.viewSize);
	x1 = getValue(state.viewX);
	x2 = x1 + wh.width;
	y1 = getValue(state.viewY);
	y2 = y1 + wh.height;

	ensureWithin(state.viewX, x1, x2, point.x);
	ensureWithin(state.viewY, y1, y2, point.y);
}

ensureWithin(v : DynamicBehaviour<double>, x1 : double, x2 : double, x : double) -> void {
	if (x < x1) {
		delta = x1 - x;
		nextDistinct(v, getValue(v) - delta);
	} else if (x > x2) {
		delta = x - x2;
		nextDistinct(v, getValue(v) + delta);
	}
}

// Finds element that satisfies given predicate and contains given position `pos`
findElementContainingPosition(state: WigiEditorState, predicate: (WigiEditorState, [int]) -> bool, pos: [int]) -> Maybe<[int]> {
	if (predicate(state, pos)) {
		Some(pos)
	} else {	// Go to parent
		if (pos == []) {	// We're at the top element already, found nothing
			None()
		} else {
			findElementContainingPosition(state, predicate, wigiParentPath(pos))
		}
	}
}

isEmptyDeletableRecursive(state : WigiEditorState, path : [int]) -> bool {
	root = getValue(state.document);
	e = getElementAtPosition(root, path);
	isComap = isWigiComap(e);
	isException = \ -> isWigiRecursive(e, [dummyWigiTable, dummyWigiTableCell, dummyWigiSpreadsheetCell, WigiVersions([]), WigiFrame(0.0, 0.0, [])]);
	areChildrenEmpty = \ -> all(map(getWigiElementChildren(e, false), isEmptyElement));

	if ((hasFixedChildCount(e) || isComap) && !isException() && areChildrenEmpty()) {
		// We should also check if we can delete this element, i.e. parent element should have non-fixed child count
		parent = getElementAtPosition(root, wigiParentPath(path));

		// There is also a special case, allow delete empty comaps under other comaps
		!hasFixedChildCount(parent) || isWigiComap(e) && isWigiComap(parent)
	} else {
		false
	}
}

hasFixedChildCount(element: WigiElement) {
	switch (element) {
		WigiRecursive(elements, type, style): {
			switch (type: WigiRecursiveType) {
				WigiTable(__, __, __, __): true;
				WigiExternalRecursive(exType, content): {
					eitherFn(
						getExternalRecursiveAPI(exType),
						\api -> {
							canDeleteFn = extractStruct(api.style, ExternalRecursiveCanDeleteChild(\__, __ -> true)).fn;
							!canDeleteFn(content, elements);
						},
						\ -> false
					)
				}
				default: wigiRecursiveChildCount(type) != -1;
			}
		}

		default: false;
	}
}

isWigiComap(e: WigiElement) -> bool {
	switch (e) {
		WigiRecursive(__, t, __):
			switch(t) {
				WigiComap(__): true;
				default: false;
			}
		default: false;
	}
}

