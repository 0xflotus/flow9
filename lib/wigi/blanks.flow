import formats/wiki/wikistyle_types;
import form;

export {
	blankPrefix = "___";
	makeBlankString(name : string) -> string;
	getExternalFormTree(style : [WikiStyle]) -> Maybe<Tree<string, Form>>;
	useExternalForm(name : string, formTree : Tree<string, Form>) -> bool;
	getExternalForm(name : string, formTree : Tree<string, Form>) -> Form;
}

makeBlankString(name : string) -> string{
	blankPrefix + name
}

getBlankName(text : string) -> string{
	if (startsWith(text, blankPrefix)) substring(text, strlen(blankPrefix), strlen(text) - 3) 
	else "";
}

getExternalFormTree(style : [WikiStyle]) -> Maybe<Tree<string, Form>> {
	fold(style, None(), \acc, s -> switch(s){
		ExternallyGeneratedForms(f): {
			tree = f();
			if (sizeTree(tree) > 0)	Some(mapTree(tree, \ws -> ws(style)))
			else acc;
		}
		default: acc;
	});
}	

getExternalForm(text : string, formTree : Tree<string, Form>) -> Form {
	blankName = getBlankName(text);
	if (blankName != "") either(lookupTree(formTree, blankName), Empty())
	else Empty();
}

useExternalForm(text : string, formTree : Tree<string, Form>) -> bool{
	getBlankName(text) != "" && sizeTree(formTree) > 0
}

