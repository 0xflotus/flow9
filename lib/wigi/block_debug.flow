// The whole file exists for debugging purposes only, really

import wigi/types;
import date;

export {
	wigiBlockToString(type : WigiBlockType) -> string;
}

wigiBlockToString(type : WigiBlockType) -> string {
	switch (type : WigiBlockType) {
		WigiPicture(url, style): toString(type);
		WigiVideo(url, style): toString(type);
		WigiAudio(url, style): toString(type);
		WigiSlider(val, minVal, maxVal, stepVal, widthVal, style): "Slider " + d2s(getValue(val));
		WigiCheckBox(val, style): "Check box " + b2s(getValue(val));
		WigiRadioButton(val, style): "Radio button" + b2s(getValue(val));
		WigiInputBox(val, w, h, style): "Input box " + getValue(val);
		WigiPositionTag(id): {
			"Position tag " + i2s(id);
		}
		WigiCustomSymbol(t, s): "Custom symbol " + toString(t);
		WigiFormdesigner(drawing, style): "Formdesigner form" + drawing;
		WigiSmartArt(content, style): "SmartArt object " + content;
		WigiFormula(formula, style): "Formula " + formula;
		WigiPageBreak(a, s): "Page break: append=" + b2s(a) + ", autoScroll=" + b2s(s);
		WigiCoach(text): "Coach: " + text;
		WigiPageName(name, disableCondition): "Page name: " + name;
		WigiNaturalMath(formula): "Math formula: " + formula;
		WigiComboBox(s, v, r, w): "Combo box: " + v;
		WigiNativeWiki(w): "Native wigi: " + w;
		WigiDotplot(r, s): "WigiDotplot range: " + r;
		WigiHistogram(r, s): "WigiHistogram range: " + r;
		WigiHistogramClasses(r, s): "WigiHistogramClasses range: " + r;
		WigiBoxplot(r, s): "WigiBoxplot range: " + r;
		WigiScatterplot(r, s): "WigiScatterplot range: " + r;
		WigiStackedBar(r, s): "WigiStackedBar range: " + r;
		WigiDatepicker(d, __): "Wigidatepicker date: " + date2string(getValue(d));
		WigiCalculator():"Calculator";
		WigiHTML(url, w, h, s):"Html resource: " + url;
		WigiReferenceDocument(t, d):"Reference document";
		WigiSetClipboardButton(title, text): "Set to clipboard button: " + title;
		WigiChoiceBlock(formula, choices): "Choice formula: " + formula;
		WigiExternalBlock(exType, content): "External block: " + exType;
		WigiFeedback(condition, description): "Feedback: " + condition;
		WigiFeedbackBlock(formula, visibilityCondition, interactiveType, choices): "Feedback formula: " + formula;
		WigiFillBlankBlock(width, caseSensitive, exactMatch, mathFeatures, answers, otherAnswer): "Fill-blank template";
	}
}
