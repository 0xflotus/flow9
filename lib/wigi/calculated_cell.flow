import arrayutils; // subtractA
import movable2; // makeWindowXY

import tropic/ttextinput;
import tropic/tropic2form;
import ui/listbox;

import wigi/editor_types;
import wigi/table_selection;
import wigi/wigi_builtins;

import ui/messagebox;
import styleform;
import wigi/wigi_button_api;

import wigi/testcaserecorder_api;

export {
	// Shows dialog where user can enter calculated formula, see case http://process.area9.dk/cases/default.asp?34549
	showCalculatedCellDialog(state: WigiEditorState, formula: string, onOK: (string) -> void) -> void;

	getExcellFunctionList() -> [string];
}

showCalculatedCellDialog(state: WigiEditorState, initialFormula: string, onOK: (string) -> void) {
	if (getValue(state.editMode)) {
		doc = getValue(state.document);

		editableFormula = make(initialFormula);
		cursorPos = getValue(state.cursor);

		// We should clear the selection and then when it starts, we'll start inserting range in the formula
		nextDistinct(state.selection, wigiClearSelection(doc));
		nextDistinct(state.selectionAnchor, None());

		// Where in editable formula the dynamic selection range is, for example:
		// formula: "=SUM(A1:B2)", start would be 5, end would be 10
		startRangePosition = ref -1;
		endRangePosition = ref -1;

		disposers = ref [];

		currentRangeStr = selectu(state.selection, \selection -> {
			switch (getRectangularSelection(selection, doc, true)) {
				Some(rectSelection): {
					// We've got some selection
					rectSelection.rangeStr
				}
				None(): {
					// We have no selection
					""
				}
			}
		});
		refArrayPush(disposers, currentRangeStr.second);

		settingFormulaNow = ref false;

		refArrayPush(disposers, subscribe2(state.selection, \selection -> {
			switch (getRectangularSelection(selection, doc, true)) {
				Some(rectSelection): {
					// We've got some selection
					curFormula = getValue(editableFormula);

					if (^startRangePosition < 0 || ^startRangePosition >= strlen(curFormula)) {
						startRangePosition := strlen(curFormula);
						endRangePosition := ^startRangePosition;
					}

					newFormula = strLeft(curFormula, ^startRangePosition) + getValue(currentRangeStr.first) + strRight(curFormula, ^endRangePosition);
					endRangePosition := ^startRangePosition + strlen(getValue(currentRangeStr.first));

					settingFormulaNow := true;
					next(editableFormula, newFormula);
					settingFormulaNow := false;

					// Select document back
					updateKeyboardFocus(^(state.focusId));
				}
				None(): {
					// We have no selection
				}
			}
		}));

		// If formula gets manually edited, we should clear selection and reset this range feature
		refArrayPush(disposers, subscribe2(editableFormula, \formula -> {
			if (!^settingFormulaNow) {
				startRangePosition := -1;
				endRangePosition := -1;

				setWigiEditorSelection(state, wigiClearSelection(doc), false);
			}
		}));

		closeFn = ref nop;
		closeMe = \ -> {
			next(state.editMode, true);
			nextDistinct(state.cursor, cursorPos);
			^closeFn();
		}

		form = {
			helpText = "Examples of formulas for calculations:\n=2 * slider + 4\n=if checkbox1 then slider + 5 else slider - 5\n=if slider &lt; 5 then \"Small\" else if slider = 5 then \"Five\" else \"Big\"";
			helpListBoxText = "Functions are compatible\nwith Excel functions\nof the same name";

			okButton = callWigiButtonWithWidth(60.0, ^withTranslation("OK"), const(true), "Ctrl+Enter", \ -> {
				closeMe();
				onOK(getValue(editableFormula));
			});

			cancelButton = callWigiButtonWithWidth(60.0, ^withTranslation("Cancel"), const(true), "esc", \ -> {
				closeMe();
			});

			focusB = make(true);
			positionB = make(strlen(initialFormula));

			editor = tropic2form(TTextInput(editableFormula, [FontSize(14.0), Multiline(true), WordWrap(true)],
					[TTextInputFocus(focusB), TTextInputPosition(positionB)], TFixed(375.0, 207.0)), emptyStylesheet);

			fillListSelected = make(-1);
			arr = getExcellFunctionList();
			listBox = ListBox(map(arr, \txt -> Text(txt, [FontSize(12.0)])), [WidthHeight(117.0, 200.0), ResetChoice(true)], fillListSelected);

			selectedName = make(-1);
			selectedFormula = make(-1);
			wigiEnv = ^(state.engine.env);
			names = concat(
				wigiEnv.cellnames |> getTreeKeys |> filterNames,
				^(getGlobalWVEnvironment().doubles) |> getTreeKeys |> filterWikiNames
			) |> uniq;
			formulas = wigiEnv.cellformulas |> getTreeKeys |> filterFormulas;

			insertText = \text, shift -> {
				curPosition = getValue(positionB);
				curFormula = getValue(editableFormula);
				newFormula = strLeft(curFormula, curPosition) + text + strRight(curFormula, curPosition);
				next(editableFormula, newFormula);
				next(focusB, true);
				next(positionB, curPosition + shift);
			}

			refArrayPush(disposers, subscribe2(fillListSelected, \v -> if (v > -1) {
				funcName = arr[v];
				insertText(funcName + "()", strlen(funcName) + 1)
			}));
			refArrayPush(disposers, subscribe2(selectedName, \v -> if (v > -1) {
				name = names[v];
				insertText(name, strlen(name))
			}));
			refArrayPush(disposers, subscribe2(selectedFormula, \v -> if (v > -1) {
				formula = formulas[v];
				insertText(formula, strlen(formula))
			}));

			definedNamesAndFormulas = \ -> {
				makeListBox = \arr1, sel -> ListBox(
					map(arr1, \txt -> Text(txt, [])),
					[WidthHeight(117.0, 76.0), ResetChoice(true)], sel
				) |> localFrame;

				Lines([
					Text("Interactives", [FontSize(14.0)]),
					makeListBox(names, selectedName),
					Fixed(0.0, 7.0),
					Text("Calculated", [FontSize(14.0)]),
					makeListBox(formulas, selectedFormula)
				])
			}

			textElements = Cols([
				Lines([Fixed(0.0, 1.0), editor]),
				Fixed(4.0, 0.0),
				localFrame(listBox),
				Fixed(4.0, 0.0),
				definedNamesAndFormulas()
			]);

			content = Lines([
				Fixed(0.0, 4.0),
				textElements,
				Cols([Text(helpText, []), Fixed(90.0, 0.0), Text(helpListBoxText, [])]),
				Fixed(0.0, 4.0),
				Group([
					Align(1.0 / 6.0, 0.0, okButton),
					Align(5.0 / 6.0, 0.0, cancelButton)
				]),
				Fixed(0.0, 4.0)
			]) |> \f -> Constructor(f, \ -> \ -> applyall(^disposers));

			// We need to make the window non-transparent for clicks, etc.
			wh = makeWH();

			opaqueContent = Access([AccessRole("dialog")],
				Group([
					Select(wh, \_wh -> {
							Interactive([
								MouseDown2(\handled, gm -> gm.inside()),
								MouseMove2(\handled, gm -> gm().inside),
								MouseUp2(\handled, gm -> gm().inside)
							],
							Rectangle(_wh.width, _wh.height, [Fill(white)]))
						}
					),
					Inspect([ISize(wh)], content)
				])
			);

			Center(makeWindowXY(make(0.0), make(0.0), opaqueContent, "Enter calculated cell"));
		}

		// Switch to the selection mode
		next(state.editMode, false);

		// And show the whole thing
		closeFn := showModal(form);
	} else {
		// We show this dialog already, so just ignore
	}
}

getExcellFunctionList() -> [string] {
	subtractA(sort(getTreeKeys(getWigiBuiltIns())),
		["+","-","*","/","=","<>","!=","<=",">=",">","<","&","[","and","or","not"])
}

localFrame(f : Form) frameAround(Border(4.0, 4.0, 4.0, 4.0, f), 1.0, black);
getNames(arr : [WigiCellName]) map(arr, \e -> toLowerCase(e.name));

filterNamesStartsWith(a : [string], text : [string]) -> [string] {
	fold(text, a, \acc, item ->
		filter(acc, \e -> !startsWith(e, item))
	)
}

filterFormulas(arr : [WigiCellName]) -> [string] {
	if (isDevMode()) getNames(arr)
	else {
		arr
		|> \a1 : [WigiCellName] -> filter(a1, \e -> length(e.namespace.names) == 0 || (!startsWith(toLowerCase(e.namespace.names[0]), "sheet")))
		|> getNames
		|> \a2 -> filterNamesStartsWith(a2, ["choice", "feedback"])
	}
}

filterNames(arr : [WigiCellName]) -> [string] {
	res = getNames(arr);
	if (isDevMode()) res
	else {
		filterNamesStartsWith(res, ["__counting", "table"])
	}
}

filterWikiNames(arr : [string]) -> [string] {
	if (isDevMode()) arr
	else {
		arr
		|> \a1 : [string] -> map(a1, \e -> ltrim2(e, "$") |> toLowerCase)
		|> \a2 : [string] -> filterNamesStartsWith(a2, ["pi", "table", "overrideconditions", "scored"])
	}
}
