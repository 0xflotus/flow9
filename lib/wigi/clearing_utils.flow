import wigi/types;

export {
	CLEARED_STYLE_KEY = "cleared";
	filterWigiParagraphExternalStyle(styles : [WigiParagraphStyle]) -> [WigiParagraphStyle];
}

filterWigiParagraphExternalStyle(styles : [WigiParagraphStyle]) -> [WigiParagraphStyle] {
	filter(styles, \extStyle ->
		switch (extStyle) {
			WigiParagraphExternalStyle(data): !getJsonBoolField(parseJson(data), CLEARED_STYLE_KEY, true);
			default: true;
		}
	);
}