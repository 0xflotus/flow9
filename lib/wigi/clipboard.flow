import wigi/editor_types;
import wigi/storage;
import serialize;
import wigi/write2wigi;
import wigi/wigi2text;
import ui/messagebox;
import ui/fontmapping;
import wigi/wigi2string;
import wigi/testcaserecorder_api;
import formats/writeparser;
import wigi/fixup;
import wigi/wigi_wiki;


export {
	//some browsers (such as IE) allow us to handle only KeyUp (not KeyDown) events for specific CTRL+Letter combinations
	// these two bools define which key handler to be used in editor (KeyUp or KeyDown) for Ctrl+letter
	handleKeyUpEvents : bool;
	handleKeyDownEvents : bool;

	innerAppClipboardKey : string = "innerAppClipboard";

	makeSystemClipboardBuffer(state : WigiEditorState, onPaste : (WigiEditorState, string) -> void) -> (()-> void);

	WigiCopyToClipboardMode ::= PreserveFormatting, PlainText, TryToSimplifyToPlainText;
		PreserveFormatting();
		PlainText();
		// Special case when we try to return plain text for ctrl+c action, when there is only WigiText without styles
		TryToSimplifyToPlainText();

	copyToClipboards(state : WigiEditorState, wigi : Maybe <WigiElement>, formatting : WigiCopyToClipboardMode) -> void;

	pasteFromClipboards(state : WigiEditorState, onPaste : (string, Maybe<ParagraphSpacing>) -> void) -> WigiSendEventResult;

	maybeGetWigiElement(content : string) -> Maybe<Pair<WigiElement, [string]>>;

	WigiSendEventResult ::= WigiSendEventResultDeferred, WigiSendEventResultImmediate, WigiSendEventResultWaitCompleted;
		WigiSendEventResultWaitCompleted(handled : bool, completed : ref () -> void);
		WigiSendEventResultDeferred(handled : bool);
		WigiSendEventResultImmediate(handled : bool);
}

// it is true when we want to use KeyUp handler in the editor
// TODO: Get rid of it? Seems it doesn't work in IE Flash which it was added for
handleKeyUpEvents = {
	flash && {
		upTypeBrowserNames = ["explorer"];
		curBrowser = getBrowser() |> toLowerCase;
		fold(upTypeBrowserNames, false, \acc, name ->{
			if (!acc){
				strIndexOf(curBrowser, name) != -1
			} else acc;
		});
	}
}

// it is true when we want to use KeyDown handler in the editor
handleKeyDownEvents = !handleKeyUpEvents;

// it is true when we need to click twice to paste
isDoublePressPaste  = {
	handleKeyUpEvents;
}

// it is true when we need to click twice to copy
isDoublePressCopy  = {
	false;
}

bufferContentStyle  = [getUniversalFont()];

// We need to filter out some characters, see case http://process.area9.dk/cases/default.asp?35327
// But we don't have to filter "line break", see case http://process.area9.dk/cases/default.asp?40471
filterTextFromNonprintables(text: string) -> string {
	if (strlen(text) == 0) text
	else doFilterText(text, 0, 0, "", \c -> isPrintableCharCode(c) || (c == 0x0A));
}

doFilterText(text : string, i : int, n : int, acc : string, checkFn : (int) -> bool) -> string {
	if ((i + n) == strlen(text))
		if (n > 0)
			acc + substring(text, i, n)
		else acc
	else if (getCharCodeAt(text, i + n) |> checkFn)
		doFilterText(text, i, n + 1, acc, checkFn)
	else if (n > 0)
		doFilterText(text, i + n + 1, 0, acc + substring(text, i, n), checkFn)
	else doFilterText(text, i + 1, 0, acc, checkFn)
}

bufferContentChange(state : WigiEditorState, text : TextInputModel) {
	if (^(state.clipboardBuffer.mode) == Pasting()) {

		filteredContent = filterTextFromNonprintables(text.content);
		if (filteredContent != text.content) {
			println("Some characters were filtered in " + text.content);
		}

		next(state.clipboardBuffer.content, filteredContent);
	}
}

clearBuffer(state : WigiEditorState) -> void {
	state.clipboardBuffer.mode := Waiting();
	next(state.clipboardBuffer.state, [Content("", bufferContentStyle), Focus(false)]);
	next(state.clipboardBuffer.content,  "");
}

firstAction = ref true;

escortBufferAction(state : WigiEditorState) -> void {
	// in case copy/paste is not completed within 1 seconds
	timer(1000, \-> {
		if (^(state.clipboardBuffer.mode) != Waiting())
			clearBuffer(state);
	});
	// clear indication of copy/paste shotrcuts when either copy or paste done for the first time
	if (^firstAction && (isDoublePressPaste || isDoublePressCopy) ) {
		timer(100, \-> {
			firstAction := false;
			arr = ref [];
			if (isDoublePressCopy)	refArrayPush(arr, "press Ctrl-C TWICE to COPY");
			if (isDoublePressPaste)	refArrayPush(arr, "press Ctrl-V TWICE to PASTE");
			disposer = warningBox2(strGlue(^arr, "\n\n"));
			timer(4000, \-> {
				resetKeyboardFocus(^(state.focusId));
				disposer();
			})
		});
	};
}


filterElementsR(children : [XmlNode], tagFilter : string) -> [XmlElement] {
	fold(children, [], \acc, v -> {
		tags = getXmlTagElements(v, tagFilter);
		if (tags != []) concat(acc, tags)
		else concat(acc, filterElementsR(getXmlChildren(v), tagFilter))
	})
}

// MS Word and Google Docs use margin-bottom attribute of <p> tag to define space after each paragraph
// wigi define ParagraphSpacing on the WigiStory level, so we use the first margin-bottom found on <p> tag
getParagraphSpacingFromHtml(htmlStr : string) -> Maybe<ParagraphSpacing> {
	pElems = filterElementsR(getXmlChildren(parseXml2(htmlStr)), "p");

	if (pElems == []) None()
	else {
		getCssProperties = \e : XmlNode -> {
			cssStr = strSplit(unescapeHtml(strReplace(getXmlAttribute(getXmlAttributes(e), "style", "default value"), "\n", "")), ";");
			filtermap(cssStr,
				\s -> { p = strSplit(s, ":"); if (length(p) == 2) Some(Pair(trim(p[0]), trim(p[1]))) else None() }
			)
		};

		// MS Word with default paragraph spacing "after" means 8pt but it does not define "margin-bottom" attribute
		// Google Docs define this attribute always, and Word Online neve provides the attribute
		// so if we have html with <p> but without "margin-bottom" it is safe to assume 8 pt
		marginBottom = findmap(pElems, \e -> findmap(getCssProperties(e), \p -> if (p.first == "margin-bottom") Some(p.second) else None()));
		pt2double = \pt, def -> if (endsWith(pt, "pt")) s2d(strLeft(pt, strlen(pt)-2)) else def;

		// wigi has by default 3 type of spacing: "single", "1.5", "double". map vaules from html to them.
		html2paragraphSpacing = \s -> ParagraphSpacing(if (s >= 6.) 6. else if (s < 6. && s >= 3.0) 3. else 0.);

		Some(html2paragraphSpacing(pt2double(either(marginBottom, "8.0pt"), 8.)))
	}
}

// PASTE for js/cpp - Ctrl-V directly, Flash - Ctrl-V (Opera, Chrome, FireFox), Ctrl-V + Ctrl-V (IE) through buffer,
pasteFromClipboards(state : WigiEditorState, onPaste : (string, Maybe<ParagraphSpacing>) -> void) -> WigiSendEventResult {
	getWigiSpacing = \ -> {
		html = getTestAwareSystemClipboard().html;
		if (html == "") None()
		else getParagraphSpacingFromHtml(html)
	}

	if (cpp || js) {
		onPaste(filterTextFromNonprintables(getGlobalClipboard()), getWigiSpacing());
		WigiSendEventResultImmediate(true);
	} else {
		if (^(state.clipboardBuffer.mode) == Waiting()) {
			escortBufferAction(state);
			state.clipboardBuffer.mode := Pasting();
			next(state.clipboardBuffer.state, [Content("", bufferContentStyle), Focus(true)]);
		}
		WigiSendEventResultImmediate(true);
	}
}

// COPY by Ctrl-C directly if it is possible,
// otherwise by Ctrl-C + Ctrl-C through buffer
copyToClipboards(state : WigiEditorState, wigiM : Maybe <WigiElement>, formatting : WigiCopyToClipboardMode) -> void {
	tryToSimplifyToPlainText = \w -> switch (w : WigiElement) {
		// copy_formulas_inside_spreadsheet.test
		WigiText(text, style): if (style == []) wigi2text(w) else wigi2string(w);
		default: wigi2string(w);
	}

	mode = switch (formatting : WigiCopyToClipboardMode) {
		PreserveFormatting(): wigi2string;
		PlainText(): wigi2text;
		TryToSimplifyToPlainText(): tryToSimplifyToPlainText;
	}
	wigi = either(wigiM, makeWigiDocument(""));
	content = eitherMap(wigiM, mode, "");

	if (!isDoublePressCopy) {
		next(state.clipboard, wigiM);
		setClipboard(content);
		customCopyFn = extractStruct(state.wigiEditorStyle, WigiEditorCustomCopy(nop3)).customCopyFn;
		customCopyFn(
			wigi,
			content,
			\newWigi -> setClipboard(mode(newWigi))
		);
	} else {
		switch (^(state.clipboardBuffer.mode)){
			Waiting() : { // first Ctrl-C
				next(state.clipboard, wigiM);
				escortBufferAction(state);
				state.clipboardBuffer.mode := Copying();
				next(state.clipboardBuffer.state, [Content(content, bufferContentStyle), Selection(0, strlen(content) ), Focus(true)]);
			}
			Copying() : { //second Ctrl-C
				// as long as selection for TextInput is deferred (case #29942) we give some time for completing
				timer(100, \-> clearBuffer(state));
			}
			Pasting() : {}
		}
	}
}

makeSystemClipboardBuffer(state : WigiEditorState, onPaste : (WigiEditorState, string) -> void) -> (()-> void) {
	if (!cpp && !js) {
		uns = selectu(state.clipboardBuffer.content, \s -> {
			if (^(state.clipboardBuffer.mode) == Pasting()) {
				onPaste(state, s);
				clearBuffer(state);
			}
		});
		// render buffer outside the window
		dispose = render(Offset( 0.0, -100.0,
			TextInput(
				[Multiline(true), Content("", bufferContentStyle), Focus(false)],
				[
					TextChange(\text -> bufferContentChange(state, text)),
					FocusOut(\-> resetKeyboardFocus(^(state.focusId)))
				],
				[StateChanger(state.clipboardBuffer.state)]
		)));

		\-> {
			uns.second();
			dispose();
		}
	} else {
		nop;
	}
}

maybeGetWigiElement(content0 : string) -> Maybe<Pair<WigiElement, [string]>> {
	content = trim2(content0, "\n \u000d\t");

	if (isWriteContent(content)) {
		write = parseWrite(content, [FontSize(20.0)]);
		if (write.stories != []) { // write content should exist
			wigi =	write2wigi(write);
			errors = if (wigi.second != [])
					concat(["XML-Wigi conversion errors:"], wigi.second)
				else [];
			Some(Pair(wigi.first, errors))
		} else None()
	} else if (isWigiContent(content)) {
		wigi = deserialize2(content, wigiFixup(), IllegalStruct());
		if (flow(wigi) != IllegalStruct()) Some(Pair(wigi, [])) else None()
	} else if (startsWith(content, "=")) {
		Some(Pair(WigiText(content, []), []))
	} else None()
}

resetKeyboardFocus(focusId : int) -> void {
	releaseKeyboardFocus();
	updateKeyboardFocus(focusId);
}
