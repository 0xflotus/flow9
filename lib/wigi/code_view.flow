import wigi/types;
import wigi/wigi_engine;
import wigi/wigi_value2form;
import ui/buttons;

export {
	code2form(engine : WigiEngine, codeView : Form, code : WigiElement, showCode : DynamicBehaviour<bool>, showResult : DynamicBehaviour<bool>, updateCursorFn : () -> void, registerToEngine : bool) -> Form;
}

code2form(engine : WigiEngine, codeView : Form, code : WigiElement, showCode : DynamicBehaviour<bool>, showResult : DynamicBehaviour<bool>, updateCursorFn : () -> void, registerToEngine : bool) -> Form {
	expr = parseWigiCode(code);
	// TODO: Come up with a better name for this guys
	value = addFormulaToEngine(engine, globalWigiNamespace, "todo", globalWigiNamespace, expr, registerToEngine, true);
	Constructor(
		Line([
			toggleControl(showCode, updateCursorFn),
			Visibleb(showCode, codeView),
			toggleControl(showResult, updateCursorFn),
			Visibleb(showResult, 
				Select(value.first, \v : WigiValue -> {
					wigiValue2Form(v, [], engine.defaultStyles) // TODO: figure out the style
				})
			)
		]),
		\ -> {
			\ -> {
				applyall(value.second)
			}
		}
	)
}


toggleControl(show : DynamicBehaviour<bool>, updateCursorFn : () -> void) -> Form {
	mkIcon = \color, ch -> {
		t = Text(ch, [FontFamily("Demi"), Fill(color), FontSize(10.0)]);
		metrics = getStaticFormSizeCached(t);
		rIcon = metrics.height / 2.0;
		correction = if (flash) 0.0 else 1.0;
		Group([
			Circle(rIcon, rIcon, rIcon, [Stroke(color), StrokeWidth(1.5)]),
			Offset(correction, if (ch == "+") 0.0 else -1.0 + correction, Align2(0.5, 0.0, t))
		])
	};

	mkButton = \c -> SimpleButton(
		mkIcon(0x333333, c),
		mkIcon(0x0066d9, c),
		"",
		\ -> {
			next(show, !getValue(show));
			updateCursorFn();
		}
	);

	Select(show, \s -> {
		mkButton(if (s) "\u2013" else "+")
	});
}

parseWigiCode(code : WigiElement) -> WigiExpr {
	// TODO: We can do better than just values
	getWigiElementValue(code);
}
