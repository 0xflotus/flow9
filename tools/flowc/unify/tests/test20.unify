import wigi;

.(.style : α30002)  c=  union WigiElement<>;

//α0   c=  union WigiElement<>;


//.(.style : α30002)  c=  α0;
