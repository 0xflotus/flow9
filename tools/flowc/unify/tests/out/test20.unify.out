α100={* .. WigiDocumentStyle<>}
α30002=[{* .. WigiDocumentStyle<>}]
